File: frontend/src/types/businessRelationship.ts
export type RelationshipType =
  | "business"
  | "reference"
  | "dependency"
  | "hierarchical"
  | "process";

export type RelationshipCardinality =
  | "one_to_one"
  | "one_to_many"
  | "many_to_one"
  | "many_to_many";

export type RelationshipApprovalStatus =
  | "draft"
  | "generated"
  | "approved"
  | "rejected";

export interface RelationshipEntitySummary {
  id: number;
  domain_id: number;
  name: string;
  description: string | null;
  classification: string;
  approval_status: string;
}

export interface BusinessRelationship {
  id: number;

  source_entity_id: number;
  target_entity_id: number;

  relationship_name: string;
  inverse_relationship_name: string | null;

  relationship_type: RelationshipType;
  cardinality: RelationshipCardinality;

  description: string | null;

  confidence: number;
  approval_status: RelationshipApprovalStatus;
  is_active: boolean;

  created_at: string;
  updated_at: string;

  source_entity: RelationshipEntitySummary;
  target_entity: RelationshipEntitySummary;
}

export interface BusinessRelationshipCreate {
  source_entity_id: number;
  target_entity_id: number;

  relationship_name: string;
  inverse_relationship_name?: string | null;

  relationship_type: RelationshipType;
  cardinality: RelationshipCardinality;

  description?: string | null;
  confidence: number;
}

export interface BusinessRelationshipUpdate {
  relationship_name?: string;
  inverse_relationship_name?: string | null;

  relationship_type?: RelationshipType;
  cardinality?: RelationshipCardinality;

  description?: string | null;
  confidence?: number;

  approval_status?: RelationshipApprovalStatus;
  is_active?: boolean;
}

export interface RelationshipSuggestion {
  source_entity_id: number;
  target_entity_id: number;

  relationship_name: string;
  inverse_relationship_name: string | null;

  relationship_type: RelationshipType;
  cardinality: RelationshipCardinality;

  description: string;
  confidence: number;
  reasons: string[];
}

File: backend/app/schemas/business_relationship.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


RelationshipType = Literal[
    "business",
    "reference",
    "dependency",
    "hierarchical",
    "process",
]


Cardinality = Literal[
    "one_to_one",
    "one_to_many",
    "many_to_one",
    "many_to_many",
]


ApprovalStatus = Literal[
    "draft",
    "generated",
    "approved",
    "rejected",
]


class BusinessRelationshipCreate(BaseModel):
    source_entity_id: int
    target_entity_id: int

    relationship_name: str = Field(
        min_length=2,
        max_length=120,
    )

    inverse_relationship_name: str | None = Field(
        default=None,
        max_length=120,
    )

    relationship_type: RelationshipType = "business"
    cardinality: Cardinality = "many_to_one"

    description: str | None = None

    confidence: int = Field(
        default=0,
        ge=0,
        le=100,
    )


class BusinessRelationshipUpdate(BaseModel):
    relationship_name: str | None = Field(
        default=None,
        min_length=2,
        max_length=120,
    )

    inverse_relationship_name: str | None = Field(
        default=None,
        max_length=120,
    )

    relationship_type: RelationshipType | None = None
    cardinality: Cardinality | None = None
    description: str | None = None

    confidence: int | None = Field(
        default=None,
        ge=0,
        le=100,
    )

    approval_status: ApprovalStatus | None = None
    is_active: bool | None = None


class RelationshipEntitySummary(BaseModel):
    id: int
    domain_id: int
    name: str
    description: str | None
    classification: str
    approval_status: str

    model_config = ConfigDict(
        from_attributes=True
    )


class BusinessRelationshipResponse(BaseModel):
    id: int

    source_entity_id: int
    target_entity_id: int

    relationship_name: str
    inverse_relationship_name: str | None

    relationship_type: str
    cardinality: str
    description: str | None

    confidence: int
    approval_status: str
    is_active: bool

    created_at: datetime
    updated_at: datetime

    source_entity: RelationshipEntitySummary
    target_entity: RelationshipEntitySummary

    model_config = ConfigDict(
        from_attributes=True
    )


class RelationshipSuggestionResponse(BaseModel):
    source_entity_id: int
    target_entity_id: int

    relationship_name: str
    inverse_relationship_name: str | None

    relationship_type: str
    cardinality: str

    description: str
    confidence: int
    reasons: list[str]
